### DISCUSSION TITLE
Zusätzliche Anzeige der Temperatur

### DISCUSSION THREAD

**Discussion by EckmarSchmitz** [tbnobody/OpenDTU#751] [Ideas]
Hallo, ich finde es sehr wichtig die Temperatur des Inverters auf dem Display anzuzeigen. Meiner Meinung nach ist das wichtiger als das Datum und die Uhrzeit. Die habe ich am Arm aber nicht die Betriebstemperatur des Inverters. In der letzten Zeile des Displays könnte im Wechsel die Temperatur mit angezeigt werden. Änderung in der Datei Display_Graphics.cpp:

```cpp 
void DisplayGraphicClass::loop()
{
    if (_display_type == DisplayType_t::None) {
        return;
    }

    if ((millis() - _lastDisplayUpdate) > _period) {
        float totalPower = 0;
        float totalYieldDay = 0;
        float totalYieldTotal = 0;
// new for temp display
	float temperature = 0;
// end
        uint8_t isprod = 0;

        for (uint8_t i = 0; i < Hoymiles.getNumInverters(); i++) {
            auto inv = Hoymiles.getInverterByPos(i);
            if (inv == nullptr) {
                continue;
            }

            if (inv->isProducing()) {
                isprod++;
            }
// new for temp display
			for (auto& c : inv->Statistics()->getChannelsByType(TYPE_INV)) {
				temperature = inv->Statistics()->getChannelFieldValue(TYPE_INV,c,FLD_T);
			}
// end
            for (auto& c : inv->Statistics()->getChannelsByType(TYPE_AC)) {
                totalPower += inv->Statistics()->getChannelFieldValue(TYPE_AC, c, FLD_PAC);
                totalYieldDay += inv->Statistics()->getChannelFieldValue(TYPE_AC, c, FLD_YD);
                totalYieldTotal += inv->Statistics()->getChannelFieldValue(TYPE_AC, c, FLD_YT);
            }
        }

        _display->clearBuffer();

        // set Contrast of the Display to raise the lifetime
        _display->setContrast(contrast);

        //=====> Logo and Lighting ==========
        //   pxMovement +x (0 - 6 px)
        uint8_t ex = enableScreensaver ? (_mExtra % 7) : 0;
        if (isprod > 0) {
            _display->drawXBMP(5 + ex, 1, 8, 17, bmp_arrow);
            if (showLogo) {
                _display->drawXBMP(_display->getWidth() - 24 + ex, 2, 16, 16, bmp_logo);
            }
        }
        //<=======================

        //=====> Actual Production ==========
        if ((totalPower > 0) && (isprod > 0)) {
            _display->setPowerSave(false);
            if (totalPower > 999) {
                snprintf(_fmtText, sizeof(_fmtText), "%2.1f kW", (totalPower / 1000));
            } else {
                snprintf(_fmtText, sizeof(_fmtText), "%3.0f W", totalPower);
            }
            printText(_fmtText, 1);
            _previousMillis = millis();
        }
        //<=======================

        //=====> Offline ===========
        else {
            printText("offline", 1);
            // check if it's time to enter power saving mode
            if (millis() - _previousMillis >= (_interval * 2)) {
                _display->setPowerSave(enablePowerSafe);
            }
        }
        //<=======================

        //=====> Today & Total Production =======
        snprintf(_fmtText, sizeof(_fmtText), "today: %4.0f Wh", totalYieldDay);
        printText(_fmtText, 2);

        snprintf(_fmtText, sizeof(_fmtText), "total: %.1f kWh", totalYieldTotal);
        printText(_fmtText, 3);
        //<=======================

        //=====> IP or Date-Time ========
        if (!(_mExtra % 10) && NetworkSettings.localIP()) {
            printText(NetworkSettings.localIP().toString().c_str(), 4);
		} 
// new for temp display
		else { 
			if (_mExtra % 5 ) {
				snprintf(_fmtText, sizeof(_fmtText), "Temperature: %.1f °C", temperature);
				printText(_fmtText,4);
			}
// end
		 	else {
            // Get current time
            time_t now = time(nullptr);
            strftime(_fmtText, sizeof(_fmtText), "%a %d.%m.%Y %H:%M", localtime(&now));
            printText(_fmtText, 4);
        	}
		}
        _display->sendBuffer();

        _dispY = 0;
        _mExtra++;
        _lastDisplayUpdate = millis();
    }
```

**Comment by Abcderfghi** [comment]
Super Idee, besteht die Möglichkeit dies in „einer Datei“ anzubieten?

**Reply by Akhilananda** [reply]
Danke, in welchem Ordner befindet sich die Datei Display_Graphics.cpp die geändert werden soll.

**Reply by oldieman** [reply]
In meinem Post vom 31. März ist die Display_Graphic.zip Datei.

**Comment by EckmarSchmitz** [comment]
Ich habe doch die Codeänderunegn/-ergänzungen schon beschrieben. Macht natürlich nur Sinn mit einem Display.

Von: Abcderfghi ***@***.***>
Gesendet: Freitag, 31. März 2023 00:36
An: tbnobody/OpenDTU ***@***.***>
Cc: Eckmar Schmitz ***@***.***>; Author ***@***.***>
Betreff: Re: [tbnobody/OpenDTU] Zusätzliche Anzeige der Temperatur (Discussion #751)


Super Idee, besteht die Möglichkeit dies in „einer Datei“ anzubieten?

**Comment by oldieman** [comment]
[Display_Graphic.zip](https://github.com/tbnobody/OpenDTU/files/11122880/Display_Graphic.zip)
Bitte ZIP-Datei entpacken und datei gegen Original austauschen. Temperatur wird dann im Wechsel zu Datum und IP-Adresse angezeigt.

**Comment by oldieman** [comment]
![dtu](https://user-images.githubusercontent.com/5185366/229194172-2bf1d65d-2eb2-45f8-ade8-a1e2b734b272.jpg)
Hier das Beispiel mit Temperatur im Gehäuse.

**Comment by anpalama1** [comment]
Ich bekomme folgende Fehler beim Compilen. Woran liegt es?

![image](https://user-images.githubusercontent.com/120574638/229292141-cc320a5c-e731-4aa7-8bae-3e40c27fbc11.png)

**Comment by oldieman** [comment]
Sorry, ich hatte das falsche Directory erwischt. War die Vorgängerversion von OpenDTU.

Sollte jetzt aber passen.

 

 

Von: anpalama1 ***@***.***> 
Gesendet: Samstag, 1. April 2023 15:36
An: tbnobody/OpenDTU ***@***.***>
Cc: oldieman ***@***.***>; Comment ***@***.***>
Betreff: Re: [tbnobody/OpenDTU] Zusätzliche Anzeige der Temperatur (Discussion #751)

 

Ich bekomme folgende Fehler beim Compilen. Woran liegt es?

 <https://user-images.githubusercontent.com/120574638/229292141-cc320a5c-e731-4aa7-8bae-3e40c27fbc11.png>

**Reply by anpalama1** [reply]
und die neue Datei?

**Comment by oldieman** [comment]
Hatte vor lauter Telefon nebenher die Datei vergessen.

 

Von: anpalama1 ***@***.***> 
Gesendet: Samstag, 1. April 2023 16:25
An: tbnobody/OpenDTU ***@***.***>
Cc: oldieman ***@***.***>; Comment ***@***.***>
Betreff: Re: [tbnobody/OpenDTU] Zusätzliche Anzeige der Temperatur (Discussion #751)

 

und die neue Datei?

**Comment by anpalama1** [comment]
fehlt immer noch ;-)

**Reply by oldieman** [reply]
[Display_Graphic.zip](https://github.com/tbnobody/OpenDTU/files/11129791/Display_Graphic.zip)
Dann geht Attachment per Email reply nicht

**Comment by Abcderfghi** [comment]
Sorry für die Frage, wie kann ich auf die Datei Display_Graphics.cpp zugreifen bzw. sie verändern?

**Reply by oldieman** [reply]
Die Datei "Display_Graphics.zip" (siehe mein letzter Post) anklicken und downloaden und dann einfach entpacken. Die darin enthaltene Display_Graphics.cpp gegen die orginal Datei austauschen. Projekt neu kompilieren.

**Reply by Abcderfghi** [reply]
Danke für die schnelle Antwort, wenn ich OpenDTU benutze finde ich keine Datei? Wo genau finde ich die Display-Datei? Wie gesagt sorry für die Frage :-)

**Reply by oldieman** [reply]
Jetzt verstehe ich die Frage nicht mehr? Schon mal mit GitHub und Visual Studio Code gearbeitet bzw. programmiert?
Wenn ja, dann sollte eigentlich klar sein was zu machen ist.
In Visual Studio Code das GitHub Project "OpenDtu" herunterladen. Dann stehen im Verzeichnis "src" die Sourcedateien zur verfügung. Auch die Datei Display_Graphic.cpp - diese dann durch den Inhalt der Display_Graghic.zip Datei aus vorherigem Post ersetzen.
Das ganze dann neu kompilieren und auf den Microcontroller laden. Das ist das gleiche wie ohne die zusätzliche Temperaturanzeige.
